lib/client.ml

```diff
@@ -251,13 +251,19 @@ let service_accepted t = function
 
 let handle_auth_failure t _ = function
   | [] -> Error "no authentication method left"
-  | xs when List.mem "publickey" xs ->
-    let pub = Hostkey.pub_of_priv t.key in
-    let met = Ssh.Pubkey (pub, None) in
-    Ok ({ t with state = Userauth_request met },
-        [ Ssh.Msg_userauth_request (t.user, service, met) ],
-        [])
-  | _ -> Error "no supported authentication methods left"
+  | xs ->
+    if List.mem "publickey" xs then
+      let pub = Hostkey.pub_of_priv t.key in
+      match t.state with
+      | Userauth_request (Ssh.Pubkey (p, None)) when Hostkey.pub_eq pub p ->
```

Comment:
Maybe a premature optimization, but we can skip comparing pubkeys since we only ever try with one key.

Reply:
I think explicitly checking for equality makes the code more readable, but on the other hand you're right...